src/stub.ts

```diff
@@ -0,0 +1,139 @@
+import sinon from 'sinon';
+import { NamespaceSchema, TypeSchema, Indexable } from 'webextensions-schema';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type BrowserOut = any;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type StubOut = any;
+
+type AnySchema = {
+  $import?: string;
+  functions?: TypeSchema[];
+  properties?: Indexable<TypeSchema>;
+  events?: TypeSchema[];
+};
+
+export class BrowserBuilder {
+  public sandbox: sinon.SinonSandbox;
+  public browser: BrowserOut;
+  public aliases: Map<string, string>;
+  public types: Map<string, TypeSchema>;
+
+  private namespaceName = '';
+
+  constructor(types: Map<string, TypeSchema>) {
+    this.sandbox = sinon.createSandbox();
+    this.aliases = new Map();
+    this.browser = {
+      sinonSandbox: this.sandbox,
+    };
+    this.types = new Map(types);
+
+    // Populate typenames without namespace prefix, in cases where no match is
+    // found for lookups using fully-qualified names
```

Comment:
I believe this is needed for cases when `ref`s don't point to full namespace names? Isn't the rule that if the type is inside the same typeschema it doesn't have a prefix and otherwise it does have one? I could see this potentially conflicting if the last part of the namespace is the same in multiple namespaces.

Reply:
Yes, this code is not bulletproof, but it fixed the issue I was seeing so I went with it! You can see the problem yourself if your comment-out that whole block https://github.com/mckenfra/webextensions-api-mock/blob/5982b253a722e487b276db03f7bba6c002077f08/src/stub.ts#L34-L42 and then do `npm run build`. You'll see warnings about references to `MutedInfo` and `SharingState` not being found in the `runtime` namespace.

We could try to work out what the 'perfect' logic for reference lookups would look like.

Reply:
Ok, I think I've cracked it. See changes in: https://github.com/stoically/webextensions-api-mock/pull/3/commits/c6eb1e608a47d9884efbbd97a4817818e828fa92

I've removed that fudge solution, and replaced it with what I'm hoping is the correct solution. The error I was seeing with the old code was due to applying the wrong `namespace` when evaluating a reference. E.g. `runtime.MessageSender` has a reference to `tabs.Tab`, which itself has a reference to `tabs.MutedInfo`. But the old code was incorrectly carrying over the `runtime` namespace and trying to lookup `runtime.MutedInfo` instead of `tabs.MutedInfo`.

The new code is also more efficient as it reuses an existing stub for a reference if the same type has previously been evaluated. E.g. `runtime.Port` is returned by `runtime.connect()`, `runtime.connectNative()` and `tabs.connect()`. WIth the new code, the same `runtime.Port` stub object is reused for all 3 occurrences.

Reply:
Wow, really elegant way of solving that one!